docs/Timelines.md

```diff
@@ -0,0 +1,149 @@
+Timelines
+---------
+
+This document covers Timelines internals. It's meant to explain how
+timelines work, to help you understand the code and what each part of
+it contributes to the overall picture.
+
+# Fake timeline: your first timeline
+
+Before timelines, Nakadi would connect to a single Kafka cluster,
+which used to be specified in the application yaml file. This was a
+static configuration loaded during boot time. Once timelines were
+introduced, in order to ease the migration process and remove special
+cases from the implementation, a fake timeline was introduced.
+
+Fake timelines exist only in memory. They are Timelines that are not
+stored in the database. Fake timelines can be seen as a fallback
+strategy for when there is no actual timeline for a given event type.
+
+Also, fake timelines use the default storage defined in the
+application configuration.
+
+# Timeline creation
+
+Timeline creation is coordinated through a series of locks and
+barriers using Zookeeper. Following we depict an example of what the
+ZK datastructure looks like at each step.
+
+## Initial state
+
+Every time a Nakadi application is launched, it tries to create the
+following ZK structure:
+
+```
+- timelines
+  + - lock                    lock for timeline versions synchronization
+  + - version: {version}      monotonically incremented long value (version of timelines configuration)
+  + - locked_et
+  + - nodes                   nakadi nodes
+    + - {node1}: {version}    Each nakadi node exposes the version used on this node
+    + - {node2}: {version}
+```
+
+In order to not override the initial structure, due to concurrency,
+each instance needs to take the lock `/nakadi/timelines/lock` before
+executing.
+
+## Start timeline creation for et_1
+
+When a new timeline creation is initiated, the first step is to
+acquire a lock to update timelines for et_1 by creating an ephemeral
```

Comment:
`et` stands for `event type`. Sometimes, we also abbreviate `timeline` into `tl`.